05-data-structures-part2.Rmd

```diff
@@ -1,245 +1,134 @@
 
-Data frames are similar to matrices, except each column can be a
-different atomic type. A data frames is the standard structure for
-storing and manipulating rectangular data sets.  Underneath the hood,
-data frames are really lists, where each element is an atomic vector,
-with the added restriction that they're all the same length.  As you
-will see, if we pull out one column of a data frame,we will have a
-vector. You will probably find that data frames are more complicated
-than vectors and other data structures we have considered, but they
-provide powerful capabilities.
-
-
-Data frames can be created manually with the `data.frame` function:
+We learned last time that the columns in a data.frame were vectors, so that our data are consistent in type throughout the column. As such, if we want to add a new column, we need to start by making a new vector:
 
 ```{r}
-df <- data.frame(id = c('a', 'b', 'c', 'd', 'e', 'f'), x = 1:6, y = c(214:219))
-df
+newCol <- c(2,3,5,12)
+cats
 ```
 
-> ## Challenge 1: Data frames {.challenge}
->
-> Try using the `length` function to query
-> your data frame `df`. Does it give the result
-> you expect?
->
-
-Each column in the data frame is simply a list element, which is why
-when you ask for the `length` of the data frame, it tells you the
-number of columns. If you actually want the number of rows, you can
-use the `nrow` function.
-
-We can add columns or rows to a data.frame using `cbind` or `rbind`
-(these are the two-dimensional equivalents of the `c` function):
-
-## Adding columns to a data frame
-
-To add a column we can use `cbind`:
+We can then add this as a column via:
 
 ```{r}
-df <- cbind(df, 6:1)
-df
+cats <- cbind(cats,  newCol)
 ```
 
-Note that R automatically names the column. We may decide to change
-the name by assigning a value using the `names` function:
+Why didn't this work? Of course, R wants to see one element in our new column for every row in the table:
 
 ```{r}
-names(df)[4] <- 'SixToOne'
+cats
+newCol <- c(4,5,8)
+cats <- cbind(cats, newCol)
+cats
 ```
 
-We can also provide a name when we add the column:
+Our new column has appeared, but it's got that ugly name at the top; let's give it something a little easier to understand:
 
 ```{r}
-df <- cbind(df, caps=LETTERS[1:6])
-df
+names(cats)[4] <- 'age'
 ```
 
-(`LETTERS` and `letters` are built-in constants.)
-
-
-## Adding rows to a data frame 
-
-To add a row we use `rbind`:
+Now how about adding rows - in this case, we saw last time that the rows of a data.frame are made of lists:
 
 ```{r}
-df <- rbind(df, list("g", 11, 42, 0, "G"))
+newRow <- list("tortoiseshell", 3.3, TRUE, 9)
+cats <- rbind(cats, newRow)
```

Comment:
Using `rbind` to append a row as a `list` to illustrate (one of) the pitfalls of factors is a bit contrived since if you constructed the new row using `data.frame` instead of `list`, R will automatically expand the factor levels to include the new one. eg:

```r
newRow <- data.frame(coat = "tortoiseshell", weight = 3.3, likes_string = TRUE, age = 9)
cats <- rbind(cats, newRow)
```